usr.bin/script/script.c

```diff
@@ -0,0 +1,481 @@
+static int	master, slave;
+static int	child, subchild;
+static size_t	outcc;
+static int	usesleep, rawout;
+static int	quiet, flush;
+static const char *fname;
+
+static int	isterm;
+static struct	termios tt;
+
+static void	done(void);
+static void	dooutput(void);
+static void	doshell(const char *);
+static void	fail(void);
+static void	finish(int);
+static void	scriptflush(int);
+static void	record(FILE *, char *, size_t, int);
+static void	consume(FILE *, off_t, char *, int);
+static void	childwait(void);
+static void	playback(FILE *);
+
+int
+main(int argc, char *argv[])
+{
+	ssize_t scc;
+	size_t cc;
+	struct termios rtt;
+	struct winsize win;
+	int aflg, pflg, ch;
+	char ibuf[BUFSIZ];
+	const char *command;
+
+	aflg = 0;
+	pflg = 0;
+	usesleep = 1;
+	rawout = 0;
+	quiet = 0;
+	flush = 0;
+	command = NULL;
+	while ((ch = getopt(argc, argv, "ac:dfpqr")) != -1)
+		switch(ch) {
+		case 'a':
+			aflg = 1;
+			break;
+		case 'c':
+			command = optarg;
+			break;
+		case 'd':
+			usesleep = 0;
+			break;
+		case 'f':
+			flush = 1;
+			break;
+		case 'p':
+			pflg = 1;
+			break;
+		case 'q':
+			quiet = 1;
+			break;
+		case 'r':
+			rawout = 1;
+			break;
+		case '?':
+		default:
+			(void)fprintf(stderr,
+			    "Usage: %s [-c <command>][-adfpqr] [file]\n",
+			    getprogname());
+			exit(EXIT_FAILURE);
+		}
+	argc -= optind;
+	argv += optind;
+
+	if (argc > 0)
+		fname = argv[0];
+	else
+		fname = "typescript";
+
+	if ((fscript = fopen(fname, pflg ? "r" : aflg ? "a" : "w")) == NULL)
+		err(EXIT_FAILURE, "fopen %s", fname);
+
+	if (pflg)
+		playback(fscript);
+
+	if (tcgetattr(STDIN_FILENO, &tt) == -1 ||
+	    ioctl(STDIN_FILENO, TIOCGWINSZ, &win) == -1) {
+		if (errno != ENOTTY) /* For debugger. */
+			err(EXIT_FAILURE, "tcgetattr/ioctl");
+		if (openpty(&master, &slave, NULL, NULL, NULL) == -1)
+			err(EXIT_FAILURE, "openpty");
+	} else {
+		if (openpty(&master, &slave, NULL, &tt, &win) == -1)
+			err(EXIT_FAILURE, "openpty");
+		isterm = 1;
+	}
+
+  if (!quiet){
+    (void)printf("Script started, output file is %s\n", fname);
+    fflush(stdout);
+  }
+
+
+	if (isterm) {
+		rtt = tt;
+		cfmakeraw(&rtt);
+		rtt.c_lflag &= ~ECHO;
+		(void)tcsetattr(STDIN_FILENO, TCSAFLUSH, &rtt);
+	}
+
+	(void)signal(SIGCHLD, finish);
+	child = fork();
+	if (child == -1) {
+		warn("fork");
+		fail();
+	}
+	if (child == 0) {
+		subchild = child = fork();
+		if (child == -1) {
+			warn("fork");
+			fail();
+		}
+		if (child)
+			dooutput();
+		else
+			doshell(command);
+	}
+
+	if (!rawout)
+		(void)fclose(fscript);
+	while ((scc = read(STDIN_FILENO, ibuf, BUFSIZ)) > 0) {
+		cc = (size_t)scc;
+		if (rawout)
+			record(fscript, ibuf, cc, 'i');
+		(void)write(master, ibuf, cc);
+	}
+	childwait();
+	return EXIT_SUCCESS;
+}
+
+static void
+childwait(void)
+{
+	sigset_t set;
+
+	sigemptyset(&set);
+	sigsuspend(&set);
+}
+
+static void
+finish(int signo)
+{
+	int die, pid, status;
+
+	die = 0;
+	while ((pid = wait4(-1, &status, WNOHANG, 0)) > 0)
```

Comment:
Could we keep original `wait3` and introduce a wrapper for `wait4` in `libc`?

Reply:
Done.